pandas/core/groupby/generic.py

```diff
@@ -1472,8 +1472,14 @@ def _apply_to_column_groupbys(self, func):
     def _fill(self, direction, limit=None):
         """Overridden method to join grouped columns in output"""
         res = super(DataFrameGroupBy, self)._fill(direction, limit=limit)
+
+        # Columns with group labels get removed by GroupBy._fill before calling
+        # into cython; if they're supposed to be part of the output, we have to
+        # add them back here.  We add all group labels which came from the
+        # input DataFrame, and are in the list of selected columns.
         output = OrderedDict(
-            (grp.name, grp.grouper) for grp in self.grouper.groupings)
+            (grp.name, grp.grouper) for grp in self.grouper.groupings
+            if grp.in_axis and grp.name in self._selected_obj)
```

Comment:
Are these conditions ever different? Thought the point of `in_axis` was to state if the grouping was a column in the selected object

Reply:
So in `df.groupby(labels)[selection].ffill()`, the first test checks whether `labels` is the name of a column in `df`, or something else (e.g. a separate series). The second test then checks whether `labels` is in `selection`.

You need both tests, as e.g. `labels` could be in `df` but not `selection`; or could not be in `df`, but have the same name as a column in `selection`.

Reply:
Hmm OK thanks. Sorry for all of the questions here but main thing I am trying to avoid is introducing new inconsistencies. Wondering why we don't see the same problem using `shift` as we do using `_fill` as they ultimately dispatch to _get_cythonized_result

Reply:
We don't have an issue with `shift` because it doesn't try to add the group labels back:
```python
>>> df = pd.DataFrame(dict(x=[1]))
>>> df.groupby('x').shift()
Empty DataFrame
Columns: []
Index: [0]
>>> df.groupby('x').ffill()
   x
0  1
```
You could argue there's some inconsistency between `shift` and `_fill` here, but it's an inconsistency that's been in pandas for a long time. If we did want to change that, would probably be a separate PR?

Reply:
> You could argue there's some inconsistency between `shift` and `_fill` here, but it's an inconsistency that's been in pandas for a long time. If we did want to change that, would probably be a separate PR?

Yea that's valid. The only thing I'm trying to avoid this is changing the output format once here and changing again in a subsequent release - it just makes for a lesser end user experience.

Do we need to add the group labels back for fill?

Reply:
Good question -- none of the other groupby transforms add labels back, so maybe we shouldn't? In general I'd prefer inconsistency to breaking changes, but we'll be breaking with 0.24 either way. Up to you?

Reply:
Totally agreed - going to be breaking in any case but would rather break once in the next major release than have to break in the next major release then break again in the subsequent one.

Can you see what breaks if you try not to add the labels back in? May be indicative of effort involved which could guide the right path forward

Reply:
If you leave the labels off, it just breaks the tests you'd expect: `test_group_fill_methods`, `test_pad_stable_sorting` and `test_pct_change` in `tests/groupby/test_transform.py`.

Reply:
Right those tests might have the wrong expectation.

Just so we are clear here is the current behavior with this PR:

```python
>>> df = pd.DataFrame([['a', 1], ['b', 2], ['a', np.nan], ['b', np.nan]], columns=['key', 'val'])
>>> df
  key  val
0   a  1.0
1   b  2.0
2   a  NaN
3   b  NaN

>>> df.groupby('key').ffill()
  key  val
0   a  1.0
1   b  2.0
2   a  1.0
3   b  2.0

>>> df.groupby('key').shift()  # or rank, cumsum, etc...
   val
0  NaN
1  NaN
2  1.0
3  2.0
```

If you get rid of the subclassed `_fill` implementation you'd get the following:

```python
>>> df.groupby('key').ffill()
   val
0  1.0
1  2.0
2  1.0
3  2.0

>>> df.groupby('key').shift()  # or rank, cumsum, etc...
   val
0  NaN
1  NaN
2  1.0
3  2.0
```

I don't think the fill methods should make an exception on the shape of the returned value when compared to any other transformation functions, hence why I'd rather you just remove `_fill` and update tests instead of the patch as is